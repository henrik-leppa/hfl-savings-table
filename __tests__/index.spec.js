import React from 'react';
import { mount } from 'enzyme';
import moment from 'moment';
import MUIDataTable from 'mui-datatables';
import Index from '../pages/index';
import renderer from 'react-test-renderer';

const savings = [
  {
    "project": 5,
    "description": "Buy cheaper paper clips",
    "start date": "2019-06-13T00:00:00+00:00",
    "category": "Office supplies",
    "responsible": "John Doe",
    "savings amount": 1000.00,
    "currency": "NULL",
    "complexity": "Medium"
  },
];

describe('Pages', () => {

  describe('Index', () => {

    const wrapper = mount(<Index savings={savings} />);

    it('renders', () => {
      expect(wrapper.find('td').at(1).text())
        .toBe(savings[0].description);
    });

    it('renders correctly', () => {
      const tree = renderer.create(<Index savings={savings} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('allows sorting by project', () => {
      const { columns } = wrapper.find(MUIDataTable).props();
      const projectColumn = columns.find(column => column.name === 'project');
      const { sort } = projectColumn.options;
      expect(sort === true || sort === undefined)
        .toBe(true);
    });

    it('allows searching only by description', () => {
      const { columns } = wrapper.find(MUIDataTable).props();
      expect(columns.every(column => {
        const { searchable } = column.options;
        if (column.name === 'description') {
          return searchable === true || searchable === undefined;
        }
        else {
          return searchable === false;
        }
      })).toBe(true);
    });

    it('searches case-insensitively', () => {
      const { caseSensitive } = wrapper.find(MUIDataTable).props().options;
      expect(caseSensitive === false || caseSensitive === undefined)
        .toBe(true);
    });

    it('renders dates in format: `DD.MM.YYYY`', () => {
      const { data } = wrapper.find(MUIDataTable).props();
      expect(data[0]['start date'])
        .toBe(moment(savings[0]['start date']).format('DD.MM.YYYY'));
    });

    it('renders "NULL" values as empty cells', () => {
      const { data } = wrapper.find(MUIDataTable).props();
      expect(savings[0].currency)
        .toBe('NULL');
      expect(data[0].currency)
        .toBe('');
    });
  });
});
