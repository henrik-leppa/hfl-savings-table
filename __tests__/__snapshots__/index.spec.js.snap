// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Index renders correctly 1`] = `
<main>
  <div
    className="MuiPaper-root MuiPaper-elevation4 MUIDataTable-paper-178 MuiPaper-rounded"
  >
    <div
      aria-label="Table Toolbar"
      className="MuiToolbar-root MuiToolbar-regular MUIDataTableToolbar-root-211 MuiToolbar-gutters"
      role="toolbar"
    >
      <div
        className="MUIDataTableToolbar-left-212"
      >
        <div
          aria-hidden="true"
          className="MUIDataTableToolbar-titleRoot-214"
        >
          <h6
            className="MuiTypography-root MUIDataTableToolbar-titleText-215 MuiTypography-h6"
          >
            HFL Savings List (by Henrik Franciscus Leppä)
          </h6>
        </div>
      </div>
      <div
        className="MUIDataTableToolbar-actions-213"
      >
        <button
          aria-describedby={null}
          aria-label="Search"
          className="MuiButtonBase-root MuiIconButton-root MUIDataTableToolbar-icon-216"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Search"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      className="MUIDataTableFilterList-root-284"
    />
    <div
      className="MUIDataTable-responsiveScroll-180"
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <table
        className="MuiTable-root MUIDataTable-tableRoot-179"
        role="grid"
        tabIndex="0"
      >
        <caption
          className="MUIDataTable-caption-182"
        >
          HFL Savings List (by Henrik Franciscus Leppä)
        </caption>
        <thead
          className="MuiTableHead-root MUIDataTableHead-main-287"
        >
          <tr
            className="MuiTableRow-root MUIDataTableHeadRow-root-290 MuiTableRow-head"
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <div
                className="MUIDataTableHeadCell-sortAction-310"
              >
                Description
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <span
                aria-describedby={null}
                className="MUIDataTableHeadCell-toolButton-313"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex={0}
                title="Sort"
              >
                <div
                  className="MUIDataTableHeadCell-data-309"
                >
                  Project
                </div>
                <div
                  className="MUIDataTableHeadCell-sortAction-310"
                >
                  <span
                    className="MuiButtonBase-root MuiTableSortLabel-root MUIDataTableHeadCell-sortLabelRoot-311"
                    onBlur={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="button"
                    tabIndex={0}
                  />
                </div>
              </span>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <div
                className="MUIDataTableHeadCell-sortAction-310"
              >
                Category
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <div
                className="MUIDataTableHeadCell-sortAction-310"
              >
                Responsible
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <div
                className="MUIDataTableHeadCell-sortAction-310"
              >
                Complexity
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <div
                className="MUIDataTableHeadCell-sortAction-310"
              >
                Savings amount
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <div
                className="MUIDataTableHeadCell-sortAction-310"
              >
                Currency
              </div>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableHeadCell-root-305 MUIDataTableHeadCell-fixedHeader-306 MuiTableCell-head"
              scope="col"
            >
              <div
                className="MUIDataTableHeadCell-sortAction-310"
              >
                Start date
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root"
        >
          <tr
            className="MuiTableRow-root MUIDataTableBodyRow-root-335 MUIDataTableBodyRow-hover-336 MuiTableRow-hover"
            id="MUIDataTableBodyRow-0"
            onClick={[Function]}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Description
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="Buy cheaper paper clips"
              >
                Buy cheaper paper clips
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Project
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title={null}
              >
                5
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Category
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="Office supplies"
              >
                Office supplies
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Responsible
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="John Doe"
              >
                John Doe
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Complexity
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="Medium"
              >
                Medium
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Savings amount
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title={null}
              >
                1000
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Currency
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title=""
              >
                
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MUIDataTableBodyCell-cellHide-340 MuiTableCell-body"
            >
              Start date
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MUIDataTableBodyCell-root-339 MuiTableCell-body"
              onClick={[Function]}
            >
              <span
                aria-describedby={null}
                className=""
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title="13.06.2019"
              >
                13.06.2019
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <table
      className="MuiTable-root"
    />
    <div
      aria-live="polite"
      className="MUIDataTable-liveAnnounce-183"
    />
  </div>
</main>
`;
